proposals/low-level-struct-improvements.md

```diff
@@ -0,0 +1,972 @@
+interfaces.
+
+```cs
+interface I1
+{
+    ref int Prop { get; }
+}
+
+struct S1 : I1
+{
+    int _field;
+    public ref int Prop => _ref field;
+
+    // When T is a struct type, like S1 this would end up returning a reference
+    // to the parameter
+    static ref int M<T>(T p) where T : I1 => ref p.Prop;
+}
+```
+
+The justification here is reasonable but it also introduces unnecessary
+friction on `struct` members that don't participate in interface invocations. 
+
+One key compatibility scenario that we have to keep in mind when approaching 
+changes here is the following:
+
+```cs
+struct S1
+{
+    ref int GetValue() => ...
+}
+
+class Example
+{
+    ref int M()
+    {
+        // Okay: this is always allowed no matter how `local` is initialized
+        S1 local = default;
+        return local.GetValue();
+    }
+}
+```
+
+This works because the safety rules for `ref` return today do not take into 
+account the lifetime of `this` (because it can't return a `ref` to internal
+state). This means that `ref` returns from a `struct` can return outside the 
+enclosing method scope except in cases where there are `ref` parameters or a 
+`ref struct` which is not *safe-to-escape* outside the enclosing method scope. 
+Hence the solution here is not as easy as allowing `ref` return of fields in 
+non-interface methods.
+
+To remove this friction the language will provide the attribute `[RefEscapes]`.
+When this attribute is applied to an instance method, instance property or 
+instance accessor of a `struct` or `ref struct` then the `this` parameter will
+be considered *ref-safe-to-escape* outside the enclosing method.
+
+This allows for greater flexibility in `struct` definitions as they can begin 
+returning `ref` to their fields. That allows for types like `FrugalList<T>`:
+
+```cs
+struct FrugalList<T>
+{
+    private T _item0;
+    private T _item1;
+    private T _item2;
+
+    public int Count = 3;
+
+    public ref T this[int index]
+    {
+        [RefEscapes]
+        get
+        {
+            switch (index)
+            {
+                case 0: return ref _item1;
+                case 1: return ref _item2;
+                case 1: return ref _item3;
+                default: throw null;
+            }
+        }
+    }
+}
+```
+
+This will naturally, by the existing rules in the span safety spec, allow
+for returning transitive fields in addition to direct fields.
+
+```cs
+struct ListWithDefault<T>
+{
+    private FrugalList<T> _list;
+    private T _default;
+
+    public ref T this[int index]
+    {
+        [RefEscapes]
+        get
+        {
+            if (index >= _list.Count)
+            {
+                return ref _default;
+            }
+
+            return ref _list[index];
+        }
+    }
+}
+```
+
+Members which contain the `[RefEscapes]` attribute cannot be used to implement 
+interface members. This would hide the lifetime nature of the member at 
```

Comment:
And overrides?

Reply:
Well, the only virtuals that exist in structs are `GetHashCode()` and `Equals(object)`.....

Reply:
Indeed.

Reply:
Given the exact structure of `object` today I don't think it's a problem to have `[RefEscapes]` on `override`. There is no method on `object` that has a `ref` return hence it can't really feed into this problem. 

Generally though the compiler does not make assumptions about `object` (or most other types). Hence we do need to consider a hypothetical corelib where a `ref` returning method is added to `object`. Lets say the following:

```cs
class Object 
{
  public virtual ref int GetIt() => ...;
}
```

Even in this case I believe it would be safe for a `struct` to `override` this method and return `ref` to its internal state. That is because there are only two ways in which this method can be invoked:

1. Through a reference to the `struct`. In that case the rules outlined here would come into effect and everything would be safe. 
1. Through a reference to `object`. At that point the `struct` is no longer on the stack because it's a boxed object. Hence the compiler interpretation of the `ref` as being *safe-to-escape* outside the enclosing method scope is correct because it's returning a `ref` to the boxed object, not a stack value. 

Even after writing that justification out I'm unsure if it should still be allowed. It feels cute and doesn't really open up any scenarios. Going to think on this a bit and see how I feel later in the day.